### DISCUSSION TITLE
Kyori's GlobalTranslator on Minestom

### DISCUSSION THREAD

**Discussion by breitwan** [Minestom/Minestom#1167] [Help]
👋. I want to use Kyori's GlobalTranslator on Minestom. 

For this I need to invoke method GlobalTranslator#render(component, locale) every time when I want to send component which returns translated component for each player (Inventory, Chat, ItemStack, Entity etc...). As for me, it's not a rational way, so I want to make something like Triton on spigot, which translates messages manipulating packets or make a static processor for the player that will pass components through itself before sending.

smth like this
```
Player.class

private static BiFunction<Player, Component, Component> componentHandler = ((player, component) -> {
    return component; // or null by default with if-else in #sendMessage
});

public void sendMessage(Component component) {
    // sendPacket (componentHandler.apply(this, component))
}

public void sendItemStack(ItemStack stack) {
    // sendPacket (componentHandler.apply(this, stack.meta.displayname))
}

PlayerInit.class
{
 Player.setComponentHandler((player, component) -> component instanceof TranslatableComponent ? GlobalTranslator.render(component, player.getLocale()) : component);
}
```

However, I was told in the discord chat that this could cause performance issues. I don't quite understand what I can do with it. (I don't want to use additional proxy or modified velocity).
It would be nice if Minestom added support for this. I think it's not good to create own extensions with your own CustomPlayer.class for something like this...

Triton (ProtocolLib with async handlers): https://github.com/tritonmc/Triton/
Velocity (using GlobalTranslator): https://github.com/PaperMC/Velocity/blob/dev/3.0.0/proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java#L310

**Comment by KrystilizeNevaDies** [comment]
I believe your best bet for now would be to do something like this:
```java
var translated = ComponentUtils.translateFor(examplePlayer, exampleComponent);
player.sendMessage(translated);
// or
ComponentUtils.translateAndSendFor(examplePlayer, exampleComponent);
```

Having a simple component manipulator isn't really what we'd like to do, having small api additions like that is not the greatest.
We might potentially want to add a component manipulation api with more complex context and manipulation options. If you wanted to propose that, a new discussion would be a great place.

**Reply by breitwan** [reply]
Oh, but what about inventory title? or entity display name? itemstack displayname?

**Reply by breitwan** [reply]
> We might potentially want to add a component manipulation api with more complex context and manipulation options. If you wanted to propose that, a new discussion would be a great place.

It would be nice, however, I'm not quite sure what to write about it. Everything I need, I wrote in this thread.

**Comment by jorisguffens** [comment]
Take a look at my library for translation: https://github.com/GufliMC/BrickI18n

It does not use the global translator in favor of namespacing, allowing you to keep your keys short and re-use translations of other extensions. It also does some extra magic to convert strings, objects, collections... in the arguments to components so your code stays clean and simple.

**Reply by breitwan** [reply]
I'm not sure that I can use it with entities and inventories.

**Reply by jorisguffens** [reply]
If you look at the main class: https://github.com/GufliMC/BrickI18n/blob/main/api/src/main/java/com/guflimc/brick/i18n/api/namespace/StandardNamespace.java

It has a "translate" method which just requires the TranslatableComponent and a Locale to translate it to a normal Component which can be used for anything.

Using the minestom specific api, a few extra useful methods are available: https://github.com/GufliMC/BrickI18n/blob/main/minestom-api/src/main/java/com/guflimc/brick/i18n/minestom/api/namespace/MinestomNamespace.java

You can just pass a Localizable (which minestom players are) to translate any TranslatableComponent to the player's locale.

All of these return a normal Component which can be used in chat, scoreboard, gui's, display names, bossbars...

**Reply by breitwan** [reply]
You probably didn't understand me. Can I create Inventory with title/Entity with displayname/ItemStack with TranslatableComponent? Once and for all occasions? If in the case of inventory I can create for each player separately, what about entities? Track all the time - just terrifying.

```
CustomInventory() {
   super(... , Component.translatable("mycustomtext"))
}
```

**Reply by jorisguffens** [reply]
It appears I didn't understand you indeed. It is also a use case I did overlook when writing my library.

I looked through the Minestom source and it looks like there is a native way to translate components in outgoing packets: https://github.com/Minestom/Minestom/blob/master/src/main/java/net/minestom/server/adventure/MinestomAdventure.java

Unfortunately I also looked through the list of supported packets and I'm afraid things like entities, inventories, items... are missing. Ofcourse my IntelliSense might be broken or I'm looking in the wrong place so you should first try it out yourself.
![image](https://user-images.githubusercontent.com/8937042/174910482-385b4b87-8108-445d-b528-74e932c7849d.png)

In case it doesn't work, you can maybe @ one of the maintainers and discuss implementing this for your use cases in a PR.
